distribution/sc-proxy-dex/src/wrapped_lp_token_merge.rs

```diff
@@ -0,0 +1,197 @@
+
+        if replic.is_some() {
+            tokens.push(replic.unwrap());
+        }
+        self.require_wrapped_lp_tokens_from_same_pair(&tokens)?;
+
+        let merged_locked_token_amount = self.merge_locked_asset_tokens_from_wrapped_lp(&tokens);
+        let attrs =
+            self.get_merged_wrapped_lp_token_attributes(&tokens, &merged_locked_token_amount);
+        let amount = self.get_merged_wrapped_lp_tokens_amount(&tokens);
+        self.burn_deposit_tokens(caller, &deposit);
+
+        self.nft_create_tokens(&wrapped_lp_token_id, &amount, &attrs);
+        let new_nonce = self.increase_wrapped_lp_token_nonce();
+
+        self.send_nft_tokens(
+            &wrapped_lp_token_id,
+            new_nonce,
+            &amount,
+            caller,
+            &opt_accept_funds_func,
+        )?;
+
+        Ok(())
+    }
+
+    fn require_deposit_empty_or_tokens_are_wrapped_lp_tokens(&self) -> SCResult<()> {
+        let wrapped_farm_token_id = self.wrapped_lp_token_id().get();
+        let caller = self.blockchain().get_caller();
+        let deposit = self.nft_deposit(&caller).get();
+        self.require_all_tokens_are_wrapped_lp_tokens(&deposit, &wrapped_farm_token_id)
+    }
+
+    fn get_wrapped_lp_tokens_from_deposit(
+        &self,
+        deposit: &[GenericTokenAmountPair<Self::BigUint>],
+    ) -> SCResult<Vec<WrappedLpToken<Self::BigUint>>> {
+        let mut result = Vec::new();
+
+        for elem in deposit.iter() {
+            result.push(WrappedLpToken {
+                token_amount: elem.clone(),
+                attributes: self
+                    .get_wrapped_lp_token_attributes(&elem.token_id, elem.token_nonce)?,
+            })
+        }
+        Ok(result)
+    }
+
+    fn require_wrapped_lp_tokens_from_same_pair(
+        &self,
+        tokens: &[WrappedLpToken<Self::BigUint>],
+    ) -> SCResult<()> {
+        let lp_token_id = tokens[0].attributes.lp_token_id.clone();
+
+        for elem in tokens.iter() {
+            require!(
+                elem.attributes.lp_token_id == lp_token_id,
+                "Lp token id differs"
+            );
+        }
+        Ok(())
+    }
+
+    fn require_all_tokens_are_wrapped_lp_tokens(
+        &self,
+        tokens: &[GenericTokenAmountPair<Self::BigUint>],
+        wrapped_lp_token_id: &TokenIdentifier,
+    ) -> SCResult<()> {
+        for elem in tokens.iter() {
+            require!(
+                &elem.token_id == wrapped_lp_token_id,
+                "Not a Wrapped Lp Token"
+            );
+        }
+        Ok(())
+    }
+
+    fn get_merged_wrapped_lp_token_attributes(
+        &self,
+        tokens: &[WrappedLpToken<Self::BigUint>],
+        merged_locked_asset_token_amount: &GenericTokenAmountPair<Self::BigUint>,
+    ) -> WrappedLpTokenAttributes<Self::BigUint> {
+        let mut lp_token_amount = 0u64.into();
+
+        tokens
+            .iter()
+            .for_each(|x| lp_token_amount += &x.attributes.lp_token_total_amount);
```

Comment:
the `lp_token_total_amount` is used with `locked_assets_invested` to calculate the locked tokens when removing the liquidity. So the total number of wrapped lp tokens generated and the `lp_token_total_amount` in the token attribute should always match. When merging wrapped_lp tokens, the `lp_token_total_amount` is the sum of all `lp_token_total_amount` from the tokens to merge. While the total number of generated wrapped_lp_tokens is the sum of all the tokens to merge. These two numbers can be different.

Reply:
indeed. the value calculated here ("lp_token_amount") was not right. fixed

Reply:
@virgil-serbanuta please double check

Reply:
I think it's fine now.